libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -613,6 +671,89 @@ string YulUtilFunctions::readFromStorage(Type const& _type, size_t _offset, bool
 	});
 }
 
+string YulUtilFunctions::dynamicReadFromStorage(Type const& _type, bool _splitFunctionTypes)
+{
+	solUnimplementedAssert(!_splitFunctionTypes, "");
+	string functionName =
+		"read_from_storage_" +
+		string(_splitFunctionTypes ? "split_" : "") +
+		"_" +
+		_type.identifier();
+	return m_functionCollector->createFunction(functionName, [&] {
+		solAssert(_type.sizeOnStack() == 1, "");
+		return Whiskers(R"(
+			function <functionName>(slot, offset) -> value {
+				value := <extract>(sload(slot), offset)
+			}
+		)")
+		("functionName", functionName)
+		("extract", dynamicExtractFromStorageValue(_type, false))
+		.render();
+	});
+}
+
+string YulUtilFunctions::updateStorageValueFunction(Type const& _type, boost::optional<unsigned> const _offset)
```

Comment:
Do you think it might be nicer to split this up into two functions, one for static offsets without an _offset argument and a dynamic one with a simple ``unsigned _offset`` argument? That'd prevent the need for the various case distinctions below and the function itself is rather small, so it'd not be too much duplication, would it? I'm not sure, what do you think?

Reply:
When I considered this, what made me use one function was the whole `_type.isValueType()` check in line 703 and the other `if` branch there, which I'd need to duplicate as well.
Additionally, there are still the `solUnimplemented` parts which will need to be added later leading to potential code duplication (or not, depending how they will be done of course)

Reply:
We could move only the whiskers part into two functions called from here?

Reply:
Yeah, nevermind, I think it's fine - once the optimizer will be able to properly deal with constant function arguments, we can remove all this anyways and simplify it to only ever use the dynamic versions.